packages/agreement-process/src/services/agreementSubmissionProcessor.ts

```diff
@@ -18,107 +17,63 @@ import {
 import { match } from "ts-pattern";
 import {
-  agreementNotFound,
   agreementNotInExpectedState,
   consumerWithNotValidEmail,
   contractAlreadyExists,
-  eServiceNotFound,
-  tenantNotFound,
 } from "../model/domain/errors.js";
 import {
   toCreateEventAgreementArchivedByUpgrade,
   toCreateEventAgreementSubmitted,
 } from "../model/domain/toEvent.js";
 import {
-  assertRequesterIsConsumer,
-  assertSubmittableState,
   matchingCertifiedAttributes,
   matchingDeclaredAttributes,
   matchingVerifiedAttributes,
   validateActiveOrPendingAgreement,
   validateSubmitOnDescriptor,
-  verifySubmissionConflictingAgreements,
 } from "../model/domain/validators.js";
 import { ApiAgreementSubmissionPayload } from "../model/types.js";
 import { UpdateAgreementSeed } from "../model/domain/models.js";
 import { agreementStateByFlags, nextState } from "./agreementStateProcessor.js";
 import { AgreementQuery } from "./readmodel/agreementQuery.js";
 import { ContractBuilder } from "./agreementContractBuilder.js";
-import { EserviceQuery } from "./readmodel/eserviceQuery.js";
 import { TenantQuery } from "./readmodel/tenantQuery.js";
 import { createStamp } from "./agreementStampUtils.js";
+import { retrieveTenant } from "./agreementService.js";
 
 export type AgremeentSubmissionResults = {
   events: Array<CreateEvent<AgreementEvent>>;
   initAgreement: Agreement;
   version: number;
 };
 
-export async function submitAgreementLogic(
-  agreementId: AgreementId,
-  payload: ApiAgreementSubmissionPayload,
-  constractBuilder: ContractBuilder,
-  eserviceQuery: EserviceQuery,
-  agreementQuery: AgreementQuery,
-  tenantQuery: TenantQuery,
-  ctx: WithLogger<AppContext>
-): Promise<[Agreement, Array<CreateEvent<AgreementEvent>>]> {
-  const logger = ctx.logger;
-  logger.info(`Submitting agreement ${agreementId}`);
-
-  const agreement = await agreementQuery.getAgreementById(agreementId);
-
-  if (!agreement) {
-    throw agreementNotFound(agreementId);
-  }
-
-  assertRequesterIsConsumer(agreement.data, ctx.authData);
-  assertSubmittableState(agreement.data.state, agreement.data.id);
-  await verifySubmissionConflictingAgreements(agreement.data, agreementQuery);
+export const processSubmitAgreement = async ({
+  agreementData,
+  eservice,
+  payload,
+  agreementQuery,
+  tenantQuery,
+  constractBuilder,
+  authData,
+  correlationId,
+}: {
+  agreementData: WithMetadata<Agreement>;
+  eservice: EService;
+  payload: ApiAgreementSubmissionPayload;
+  agreementQuery: AgreementQuery;
+  tenantQuery: TenantQuery;
+  constractBuilder: ContractBuilder;
+  authData: AuthData;
+  correlationId: string;
+}): Promise<[Agreement, Array<CreateEvent<AgreementEvent>>]> => {
+  const agreement = agreementData.data;
 
-  const eservice = await eserviceQuery.getEServiceById(
-    agreement.data.eserviceId
-  );
-  if (!eservice) {
-    throw eServiceNotFound(agreement.data.eserviceId);
-  }
+  const consumer = await retrieveTenant(agreement.consumerId, tenantQuery);
```

Comment:
What's the reason to retrieve from readmodel here in the processor rather than in `agreementService.ts` file? I see that you retrieve Agreement and ESeervice in `agreementService.ts` before calling this function, while you retrieve Tenants (consumer & producer) here.

Reply:
> What's the reason to retrieve from readmodel here in the processor rather than in `agreementService.ts` file? I see that you retrieve Agreement and ESeervice in `agreementService.ts` before calling this function, while you retrieve Tenants (consumer & producer) here.

Here I made a mistake, `retrieveTenant` should be called after `validateSubmitOnDescriptor` because there's no need to retrieve the tenant if that function throws.

`Agreement` is needed in `agreementService.ts` for various validations before calling the processor while `Eservice` is used as soon as we enter the processor and we could retrieve it in it too but in order to do so I would need to pass to it the `EserviceQuery` parameter, so, in that case, I preferred to retrieve it just before calling the processor.

What do you think about this?

Reply:
If I understood correctly you retrieve tenants later in the code (in the processor file and not in the main `agreementService.ts`) so that the order of validations stays the same as before. If that's the case, it works for me - even though for consistency I would love to have all readmodel retrievals in the main service file where most of them are already :)

Reply:
I had the same curiosity. 
I agree on the preference of delaying retrievals until they are actually needed, but `verifyConsumerDoesNotActivatePending` and `assertActivableState` are business logic checks and are in the service, while `validateActivationOnDescriptor` (and tenant retrieve) are in the process function.
In this case I don't see a clear pattern.

No strong opinions, I just would have expected them all in one function or the other

Reply:
I reorganized the code based on your feedbacks, let me know what you think @ecamellini @galales